v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/SpannerTransactionWriterDoFn.java

```diff
@@ -172,25 +176,35 @@ class SpannerTransactionWriterDoFn extends DoFn<FailsafeElement<String, String>,
   private transient AtomicBoolean keepWatchdogRunning;
   private transient Thread watchdogThread;
 
+  private final boolean usesSeparateShadowTableDb;
+
   SpannerTransactionWriterDoFn(
       SpannerConfig spannerConfig,
+      SpannerConfig shadowTableSpannerConfig,
       PCollectionView<Ddl> ddlView,
+      PCollectionView<Ddl> shadowTableDdlView,
       String shadowTablePrefix,
       String sourceType,
       Boolean isRegularRunMode) {
     Preconditions.checkNotNull(spannerConfig);
     this.spannerConfig = spannerConfig;
+    this.shadowTableSpannerConfig = shadowTableSpannerConfig;
     this.ddlView = ddlView;
+    this.shadowTableDdlView = shadowTableDdlView;
     this.shadowTablePrefix =
         (shadowTablePrefix.endsWith("_")) ? shadowTablePrefix : shadowTablePrefix + "_";
     this.sourceType = sourceType;
     this.isRegularRunMode = isRegularRunMode;
+    this.usesSeparateShadowTableDb =
+        !(spannerConfig.getInstanceId().equals(shadowTableSpannerConfig.getInstanceId())
+            && spannerConfig.getDatabaseId().equals(shadowTableSpannerConfig.getDatabaseId()));
   }
 
   /** Setup function connects to Cloud Spanner. */
   @Setup
   public void setup() {
     spannerAccessor = SpannerAccessor.getOrCreate(spannerConfig);
+    shadowTableSpannerAccessor = SpannerAccessor.getOrCreate(shadowTableSpannerConfig);
```

Comment:
if shadowTable is not different, we should probably not be creating an additional spannerConfig here